[WebDriver][BiDi] Implement the browser module

yury-s opened the pull request:
#### 55e2a080047d0ddb30ce789103d3faa3dbcbf0f1
<pre>
[WebDriver][BiDi] Implement the browser module
<a href="https://bugs.webkit.org/show_bug.cgi?id=281941">https://bugs.webkit.org/show_bug.cgi?id=281941</a>

Reviewed by BJ Burg.

Basic implementation of UserContext management methods. Each user context
is a pair of a WebsiteDataStore and a WebProcessPool.

* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp: Added. Browser
module logic is extracted into its own module for better readability and
encapsulation of the implementation details. The agent maintains the list
of all UserContexts, in the future they will be used to create pages (aka
BrowsingContexts).

(WebKit::BidiBrowserAgent::BidiBrowserAgent):
(WebKit::BidiBrowserAgent::close):
(WebKit::BidiBrowserAgent::createUserContext):
(WebKit::BidiBrowserAgent::getUserContexts):
(WebKit::BidiBrowserAgent::removeUserContext):
(WebKit::BidiBrowserAgent::platformCreateUserContext):
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.h: Added. An actual
implementation of the user context. It is mostly WebsiteDataStore that
provides user data isolation. After decoupling of the network process
from WebProcessPool in <a href="https://commits.webkit.org/229885@main">https://commits.webkit.org/229885@main</a>, we might
not need a separate process pool per user context for cookie store
isolatio. But we still need it for e.g. geo location overrides using
WebGeolocationManagerProxy which is a supplement to WebProcessPool.

* Source/WebKit/UIProcess/Automation/BidiUserContext.cpp: Added.
(WebKit::BidiUserContext::BidiUserContext):
* Source/WebKit/UIProcess/Automation/BidiUserContext.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/glib/BidiBrowserAgentGlib.cpp: Added.
(WebKit::BidiBrowserAgent::platformCreateUserContext):
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/293942@main">https://commits.webkit.org/293942@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9f1d4f6114051829672238c7fc4e2a5bc35d26e5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100393 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20045 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10344 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105529 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50981 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20351 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28519 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103400 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15586 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90684 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56793 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15404 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50349 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/85318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8766 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107884 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27511 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20182 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85392 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27874 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86884 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84930 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21591 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29614 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/7343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21431 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27446 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32687 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27257 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30575 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28815 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -0,0 +1,145 @@
+    return userContextID;
+}
+
+Inspector::CommandResult<Ref<JSON::ArrayOf<UserContextInfo>>>  BidiBrowserAgent::getUserContexts()
```
Nit: extra space

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.h:
```diff
@@ -0,0 +1,83 @@
+    Inspector::CommandResult<void> removeUserContext(const String& userContext) override;
+
+private:
+    class UserContext {
```
I strongly prefer to make this its own top level class in a separate file like `BidiUserContext.h`

yury-s replied:
Done. It's better that way.

burg commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.h:
```diff
@@ -0,0 +1,83 @@
+        ~UserContext();
+
+        Ref<WebsiteDataStore> dataStore;
+        Ref<WebProcessPool> processPool;
```
What does it (eventually) use the process pool for?

yury-s replied:
This is going to be used e.g. for `emulation.setGeolocationOverride` implementation to access `processPool->supplement<WebGeolocationManagerProxy>();`. But as discussed in Slack, many things have changed since we did it that way in Playwright and we might be able to avoid storing an explicit reference.

burg commented on Source/WebKit/UIProcess/Automation/glib/BidiBrowserAgentGlib.cpp:
```diff
@@ -0,0 +1,72 @@
+    GRefPtr<WebKitWebsiteDataManager> dataManager = adoptGRef(webkit_website_data_manager_new_ephemeral());
+#endif
+
+    GRefPtr<WebKitWebContext> context = adoptGRef(WEBKIT_WEB_CONTEXT(g_object_new(WEBKIT_TYPE_WEB_CONTEXT,
```
Based on our conversation in Slack, it may not be necessary to create a new process pool in order to get isolated networking sessions for each user context.

yury-s replied:
Adding a note in the commit message.

burg commented on Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
```diff
@@ -15,15 +15,48 @@
+            "description": "Creates a user context.",
+            "spec": "https://w3c.github.io/webdriver-bidi/#command-browser-createUserContext",
+            "returns": [
+                { "name": "userContext", "$ref": "BidiBrowser.UserContext", "description": "Id of created user context." }
```
Nit: Id -> Identifier

yury-s committed:
[WebDriver][BiDi] Implement the browser module
https://bugs.webkit.org/show_bug.cgi?id=281941

Reviewed by BJ Burg.

Basic implementation of UserContext management methods. Each user context
is a pair of a WebsiteDataStore and a WebProcessPool.

* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp: Added. Browser
module logic is extracted into its own module for better readability and
encapsulation of the implementation details. The agent maintains the list
of all UserContexts, in the future they will be used to create pages (aka
BrowsingContexts).

(WebKit::BidiBrowserAgent::BidiBrowserAgent):
(WebKit::BidiBrowserAgent::close):
(WebKit::BidiBrowserAgent::createUserContext):
(WebKit::BidiBrowserAgent::getUserContexts):
(WebKit::BidiBrowserAgent::removeUserContext):
(WebKit::BidiBrowserAgent::platformCreateUserContext):
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.h: Added. An actual
implementation of the user context. It is mostly WebsiteDataStore that
provides user data isolation. After decoupling of the network process
from WebProcessPool in https://commits.webkit.org/229885@main, we might
not need a separate process pool per user context for cookie store
isolatio. But we still need it for e.g. geo location overrides using
WebGeolocationManagerProxy which is a supplement to WebProcessPool.

* Source/WebKit/UIProcess/Automation/BidiUserContext.cpp: Added.
(WebKit::BidiUserContext::BidiUserContext):
* Source/WebKit/UIProcess/Automation/BidiUserContext.h: Added.
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp:
(WebKit::WebDriverBidiProcessor::WebDriverBidiProcessor):
* Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.h:
* Source/WebKit/UIProcess/Automation/glib/BidiBrowserAgentGlib.cpp: Added.
(WebKit::BidiBrowserAgent::platformCreateUserContext):
* Source/WebKit/UIProcess/Automation/protocol/BidiBrowser.json:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/293942@main
